### ISSUE TITLE
[Feature] pipe如何输出scores

### ISSUE THREAD

**Issue by KooSung** [InternLM/lmdeploy#2448]
### Motivation

下面是使用torch原生推理InternVL2-8B获取第一个token为A-Z字母的对应的26个概率list。如果使用pipe应该怎么实现呢？logprobs似乎没法实现
```python
alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
indices = []
for letter in alphabet:
    idx = tokenizer(letter)['input_ids'][1]
    indices.append(idx)

generation_config = dict(
    num_beams=1,
    max_new_tokens=1,
    do_sample=False,
    output_scores=True,
    return_dict_in_generate=True
)

response = model.chat(tokenizer, pixel_values, question, generation_config)
scores = response.scores[0]
scores = nn.Softmax(dim=-1)(scores).detach().cpu().numpy()
score = np.round(scores[:, indices][0], 4)
```

```
from lmdeploy import pipeline
from lmdeploy.vl import load_image
from lmdeploy import pipeline, GenerationConfig, TurbomindEngineConfig
import nest_asyncio
nest_asyncio.apply()

pipe = pipeline('model/InternVL2-8B')
image = load_image('demo.jpg')
gen_config = GenerationConfig(top_p=0.8,
                              top_k=40,
                              temperature=0.8,
                              max_new_tokens=1024,
                              logprobs=26)
response = pipe((f'describe this image', image), gen_config=gen_config)
```

```
Response(
text='The image depicts a table with various items on it, including a black cellphone with a cigarette in it, a box of ammunition labeled "M87 9mm", and a plastic container with more ammunition. There are also some other items such as a pair of shoes and a backpack. The table and items appear to be placed in an outdoor setting, possibly on a skateboard ramp or similar structure. The overall scene seems to suggest a casual, possibly recreational or personal setting.',  
generate_token_len=97,  
input_token_len=3381, 
session_id=0, 
finish_reason='stop',  
token_ids=[918, 2321, 2323, 30254, 395, 2115, 579, 5417, 3750, 519, 563, 328, 2832, 395, 3851, 56163, 579, 395, 35113, 435, 563, 328, 395, 3905, 446, 36597, 30093, 461, 307, 4114, 262, 305, 3977, 628, 454, 395, 12341, 5634, 579, 937, 36597, 281, 2779, 657, 1225, 1187, 1148, 3750, 1893, 569, 395, 6865, 446, 15418, 454, 395, 33236, 281, 707, 2115, 454, 3750, 5153, 442, 517, 9229, 435, 589, 15920, 6399, 328, 10915, 519, 395, 42425, 2636, 22572, 607, 4590, 6098, 281, 707, 8216, 6262, 5136, 442, 4348, 395, 16467, 328, 10915, 35952, 607, 4505, 6399, 281],  
logprobs=[{918: 0.0}, {2321: 0.0}, {5092: -0.3205167353153229, 2323: -1.2938023805618286}, {30254: 0.0}, {395: 0.0}, {6262: -1.202095627784729, 1847: -1.54282808303833, 2115: -1.6608434915542603, 22643: -1.89496648311615, 3428: -1.928551197052002}, {579: -0.0898083969950676, 3643: -2.454646110534668}, {5417: -0.5344875454902649, 3963: -0.8818157315254211}, {3750: -0.34446704387664795, 6325: -1.2330511808395386}, {519: -0.3840368092060089, 328: -1.6848713159561157, 17252: -2.0142133235931396}, {563: 0.0}, {328: -0.5174750685691833, 281: -0.9063986539840698}, {2832: 0.0}, {395: -0.2040911614894867, 1593: -1.6894997358322144}, {3929: -0.8219655752182007, 3851: -1.6514713764190674, 21752: -1.7980587482452393, 3905: -2.1534931659698486, 35113: -2.4422202110290527}, {21752: -1.2979393005371094, 56163: -1.6831779479980469, 14485: -2.0194091796875, 18441: -2.4249839782714844, 2943: -2.6544721126556396, 1781: -3.1141700744628906, 4702: -3.1294476985931396, 1286: -3.1971817016601562, 6526: -3.3694074153900146, 3831: -3.440295934677124, 15207: -3.5941314697265625, 12341: -3.627227544784546}, {328: -0.6286396980285645, 579: -0.7621045112609863}, {395: -0.2369438260793686, 16338: -1.5560660362243652}, {35113: -0.39830347895622253, 13168: -1.3862910270690918, 4316: -2.544135570526123}, {19843: -0.5642545223236084, 13168: -1.5113065242767334, 435: -2.036891460418701, 16338: -2.523719310760498}, {1326: -0.6022438406944275, 563: -1.4520896673202515, 410: -1.5217115879058838}, {328: 0.0}, {395: -0.18746504187583923, 1187: -2.9306252002716064, 3963: -3.2042040824890137, 6976: -3.25248646736145, 1028: -3.2620651721954346}, {3929: -0.636863112449646, 3905: -0.833052933216095, 4316: -3.315089464187622}, {446: 8.20042167504198e-09}, {36597: -0.7211176156997681, 33168: -0.9341760873794556, 262: -2.1129403114318848}, {30093: -0.5964093208312988, 328: -0.8002553582191467}, {461: 0.0}, {307: -0.9265238046646118, 317: -1.6707311868667603, 6744: -2.092080593109131, 309: -2.3185641765594482, 41295: -2.8605809211730957, 9516: -3.0255038738250732, 321: -3.100525379180908, 305: -3.1390576362609863}, {305: -1.3047099113464355, 2040: -1.4937928915023804, 22362: -1.9403338432312012, 314: -2.5406947135925293, 4114: -3.0430750846862793, 16117: -3.2461047172546387, 319: -3.265887498855591, 21944: -3.2741751670837402, 3821: -3.565612316131592, 972: -3.66853666305542, 285: -3.689422130584717, 1098: -3.8889193534851074, 904: -3.9175503253936768}, {281: -0.7758625149726868, 262: -0.8980758786201477, 311: -2.833296537399292, 628: -3.2569682598114014, 1481: -3.3519158363342285}, {321: -0.4494590759277344, 305: -1.0160369873046875}, {3977: 0.0}, {1481: -0.7298322319984436, 628: -1.366146445274353, 575: -2.183372974395752, 278: -2.511265277862549, 23784: -2.6722607612609863}, {454: -0.6138325929641724, 395: -0.7792988419532776}, {395: -0.24168576300144196, 3963: -2.145017385482788, 2595: -2.326558828353882}, {14555: -1.369643211364746, 3929: -1.725600242614746, 25677: -2.2949514389038086, 3851: -2.390681266784668, 4316: -2.70450496673584, 12341: -2.8056440353393555, 5634: -2.850226402282715, 5767: -3.1394147872924805, 26737: -3.2323904037475586, 41601: -3.297488212585449, 2773: -3.3095922470092773, 3905: -3.3836886882781982}, {34706: -1.2269995212554932, 5634: -1.6110249757766724, 14555: -1.8931028842926025, 17858: -2.0712149143218994, 36597: -2.381746530532837, 3905: -2.6160316467285156, 11988: -2.7329294681549072}, {10100: -0.801814079284668, 10336: -1.1359014511108398, 579: -1.4681425094604492}, {937: -0.4273316562175751, 5268: -1.7816176414489746, 3963: -2.2510504722595215, 33168: -2.6023573875427246}, {36597: 0.0}, {281: 0.0}, {2779: 0.0}, {505: -0.2855437099933624, 657: -1.392736554145813}, {1225: 0.0}, {1187: -0.9365928769111633, 395: -1.0970122814178467, 1524: -1.7335747480392456, 3963: -2.7855613231658936, 17572: -3.3290908336639404}, {33168: -1.739129662513733, 4505: -1.8905349969863892, 1148: -1.8947349786758423, 20354: -2.2391676902770996, 14485: -2.535661220550537, 4446: -2.5635390281677246, 3750: -3.1904730796813965, 15418: -3.2685370445251465, 15821: -3.3512625694274902, 6325: -3.3704276084899902, 17572: -3.481088161468506, 2582: -3.51804518699646, 3851: -3.6263327598571777, 36884: -3.7084670066833496}, {3750: -0.6210796236991882, 6325: -1.3079731464385986, 4505: -2.33245587348938, 2773: -2.3516550064086914}, {328: -1.212571620941162, 1217: -1.227918028831482, 1893: -1.4457335472106934, 519: -2.2899985313415527, 454: -2.619668483734131}, {569: 0.0}, {395: 0.0}, {6865: -1.190872311592102, 33236: -1.2279930114746094, 3851: -1.4878579378128052, 9100: -2.1953353881835938, 2773: -2.7182960510253906}, {446: 0.0}, {15418: -0.3892870843410492, 20969: -1.5231332778930664, 3851: -2.259213447570801}, {454: -0.5962911248207092, 328: -0.8004002571105957}, {395: -0.1501782089471817, 1272: -1.9700825214385986}, {33236: -0.5297783613204956, 9100: -1.271157145500183, 3851: -2.034409284591675}, {281: -0.36863428354263306, 435: -1.351334571838379, 328: -3.007256269454956}, {707: 0.0}, {2115: -1.1386960744857788, 6262: -1.2607091665267944, 6399: -1.9692432880401611, 1847: -2.0657551288604736, 4160: -2.38226056098938, 2321: -3.2785122394561768}, {8088: -0.6031427979469299, 505: -1.1205530166625977, 454: -2.065056562423706}, {14723: -0.8039295673370361, 410: -0.8721973896026611, 3750: -2.006924867630005}, {5153: -0.6017714738845825, 4348: -1.278117299079895, 657: -1.7510100603103638}, {442: 0.0}, {517: 0.0}, {435: -1.281785249710083, 9229: -1.3302509784698486, 519: -1.425828218460083, 33612: -1.524461030960083}, {519: -0.5971465110778809, 435: -0.79935222864151}, {589: 8.20042167504198e-09}, {15920: 0.0}, {6399: -0.5640003681182861, 607: -0.8414814472198486}, {328: -0.21173615753650665, 281: -1.6564152240753174}, {10915: 0.0}, {519: -0.5339396595954895, 395: -0.8825915455818176}, {395: 0.0}, {42425: -0.7522440552711487, 13566: -0.8303423523902893, 14316: -2.3773927688598633}, {2636: -0.1433628797531128, 37448: -2.0132014751434326}, {22572: -0.40312135219573975, 6271: -1.1033166646957397}, {281: -0.612126350402832, 607: -0.781315803527832}, {4590: -0.32855963706970215, 395: -1.2728228569030762}, {6098: 0.0}, {281: 0.0}, {92542: -0.3587445914745331, 707: -1.1991604566574097}, {2321: -1.438124656677246, 1847: -1.4804333448410034, 56163: -1.7109020948410034, 8216: -2.184206962585449, 6262: -2.6496152877807617, 9489: -2.929171562194824, 35113: -3.136946678161621, 5404: -3.238295316696167, 4160: -3.350886344909668}, {6262: 0.0}, {13371: -0.5373390913009644, 5136: -1.7905892133712769, 6846: -1.8806045055389404, 6022: -2.339878559112549}, {442: -0.3986336290836334, 16467: -1.1124168634414673}, {4348: -0.7844711542129517, 517: -0.846784234046936, 21270: -2.164151668548584}, {395: -0.3504520654678345, 589: -1.6549830436706543, 560: -2.2582316398620605}, {16467: 0.0}, {328: -0.6484090685844421, 607: -0.7399808168411255}, {10915: -0.15546439588069916, 8498: -1.9380645751953125}, {3329: -0.5836606025695801, 35952: -1.68803071975708, 15794: -1.8157237768173218, 52518: -2.3586888313293457}, {4736: -0.750227689743042, 607: -1.18746817111969, 6399: -1.5017114877700806}, {15794: -1.4668854475021362, 15920: -1.754494547843933, 3751: -2.293633460998535, 8725: -2.34719181060791, 34932: -2.600327253341675, 39961: -2.880997657775879, 52518: -2.9738359451293945, 4505: -3.0930335521698, 42425: -3.2121474742889404, 1115: -3.216526985168457, 41473: -3.2385261058807373, 3329: -3.4942710399627686, 38472: -3.769200086593628}, {6399: -0.475861132144928, 4736: -0.9711422920227051}, {281: 0.0}])  
```

**Comment by irexyc** [comment]
目前还不能在生成阶段返回全部的 logits (vocab_size)。

有另外一种方式有一些限制，不知道符不符合你的需求。Example 如下：

```python
from lmdeploy import pipeline, TurbomindEngineConfig
from lmdeploy.vl import load_image
pipe = pipeline('/nvme/shared/internlm-xcomposer2-7b/', backend_config=TurbomindEngineConfig(cache_max_entry_count=0.5))

# logits
image = load_image('https://raw.githubusercontent.com/open-mmlab/mmdeploy/main/tests/data/tiger.jpeg')
inputs = pipe.prepare_inputs(('describe this image', image))
input_ids = inputs['input_ids']
embeddings = inputs['input_embeddings']
embedding_ranges = inputs['input_embedding_ranges']

# get first generated token score.
# logits = pipe.get_logits(tmp_ids, embeddings, embedding_ranges)


out = pipe(('describe this image', image))
output_ids = out.token_ids

from copy import deepcopy
tmp_ids = deepcopy(input_ids)
for idx in output_ids:
    tmp_ids[0].append(idx)
    logits = pipe.get_logits(tmp_ids, embeddings, embedding_ranges)
```

限制的话，就是没办法并发。需要算两遍，第一遍获取生成的token_ids，第二遍一次计算下一个token的概率。另外可以看下get_logits内部的逻辑，这里是可以通过设置start, end来避免重复计算的。
